Support Linux DRM backend

Bennctu opened the pull request:
Currently, using the DRM legacy interface is suitable for Mado's backend because we only need basic window display without advanced features.

<img src="https://github.com/user-attachments/assets/ec9f0ccb-e788-4f69-9896-85826c2d5c0a" width="500">

jserv commented on backend/drm.c:
```diff
@@ -0,0 +1,345 @@
+}
+
+
+/* Register the Linux framebuffer backend */
```
Refine the message.

jserv commented on backend/drm.c:
```diff
@@ -0,0 +1,345 @@
+    return false;
+}
+
+static int twin_vt_open(int vt_num)
```
The VT related code should be moved to `backend/linux_vt.c`, which is shared by both Linux framebuffer and DRM backends.

jserv commented:
> Currently, using the DRM legacy interface is suitable for Mado's backend because we only need basic window display without advanced features.

What compatibility risks exist when using DRM legacy interfaces with newer Linux kernels? Could you verify the compatibility status of libdrm across different kernel versions?

What prevents the DRM backend from rendering content across the full screen resolution?

jserv commented:
@a1091150, You might check DRM backend for your environment as well.

jserv commented on Makefile:
```diff
@@ -110,6 +110,14 @@ libtwin.a_files-y += backend/fbdev.c
+BACKEND = drm
+libtwin.a_files-y += backend/drm.c
+libtwin.a_files-y += backend/linux_input.c
+libtwin.a_cflags-y += $(shell pkg-config --cflags libdrm)
+TARGET_LIBS += $(shell pkg-config --libs libdrm)
```
Mention the build dependency in the top-level documentation.

p96114175 commented on backend/drm.c:
```diff
@@ -0,0 +1,345 @@
+
+    return true;
+
+bail_crtc:
```
I'm confused about this. It should be as below.
bail_fb:
    drmModeRmFB(tx->drm_dri_fd, tx->fb_id);
    munmap(tx->fb_base, tx->fb_len);
bail_crtc:
    drmModeFreeCrtc(CRTC(tx));
    drmModeFreeConnector(CONN(tx));

jserv reviewed (changes requested):
Rebase the latest `main` branch and resolve conflicts.

jserv commented on README.md:
```diff
@@ -83,6 +83,9 @@ For SDL backend, install the [SDL2 library](https://www.libsdl.org/).
 * macOS: `brew install sdl2`
 * Ubuntu Linux / Debian: `sudo apt install libsdl2-dev`
 
+For DRM backend, install the [libdrm](https://cgit.freedesktop.org/mesa/drm/).
```
Change the hyperlink to https://gitlab.freedesktop.org/mesa/drm

Bennctu commented:
> > Currently, using the DRM legacy interface is suitable for Mado's backend because we only need basic window display without advanced features.
> 
> What compatibility risks exist when using DRM legacy interfaces with newer Linux kernels? Could you verify the compatibility status of libdrm across different kernel versions?
> 
> What prevents the DRM backend from rendering content across the full screen resolution?

I need more time to research and adjust.

jserv commented on backend/drm.c:
```diff
@@ -0,0 +1,337 @@
+    return false;
+}
+
+/* Virtual terminal part is same as fbdev */
```
Decouple this to another pull request to share the Linux VT manipulation.

jserv commented on backend/drm.c:
```diff
@@ -0,0 +1,337 @@
+{
+    *resources = drmModeGetResources(fd);
+    if (*resources == NULL) {
+        log_error("drmModeGetResources failed");
```
Make the error messages consistent, starting with "Fail to ..."

shengwen-tw commented on configs/Kconfig:
```diff
@@ -12,6 +12,10 @@ config BACKEND_FBDEV
     select CURSOR
 
+config BACKEND_DRM
+    bool "DRM support"
```
Consider changing to `Linux DRM support` to avoid confusion.

shengwen-tw commented on backend/drm.c:
```diff
@@ -0,0 +1,337 @@
+    uint32_t fb_id;
+    uint32_t *fb_base;
+    size_t fb_len;
+    // int display_width;
```
Leave no dead code.

shengwen-tw commented on backend/drm.c:
```diff
@@ -0,0 +1,337 @@
+    twin_coord_t width = right - left;
+    int line_length = 4 * tx->width;
+    off_t off = line_length * top + 4 * left;
+    unsigned char *dest = (unsigned char *) tx->fb_base + off;
```
Replace `unsigned char` with `uint8_t` for consistency and use `uintptr_t` for pointer arithmetic. That is:
```C
uint8_t *dest = (uint8_t *)((uintptr_t) tx->fb_base + off);
```

shengwen-tw commented on backend/drm.c:
```diff
@@ -0,0 +1,337 @@
+    return NULL;
+}
+
+
```
Remove one empty line here.

shengwen-tw commented on backend/drm.c:
```diff
@@ -0,0 +1,337 @@
+    free(ctx->priv);
+    free(ctx);
+}
+
```
Same as above.

jserv commented:
Check the DRM implementation for reference: https://github.com/zlgopen/awtk-linux-fb/blob/master/awtk-port/lcd_linux/lcd_linux_drm.c

jserv reviewed (changes requested):
Rebase the latest `main` branch and utilize `backend/linux_vt.h`.

jserv commented on backend/drm.c:
```diff
@@ -0,0 +1,300 @@
+
+/* Register the Linux DRM backend */
+
+const twin_backend_t g_twin_backend = {
```
You should add `poll` operation as it is required in recent `main` branch.

Bennctu replied:
As the original implementation, we creates another thread to handle the events in function [`twin_linux_evdev_thread`](https://github.com/sysprog21/mado/blob/35ddf50ac4c7fe61b8b3c963f80dabceeb7783fe/backend/linux_input.c#L301). So do you suggest refactoring [linux_input.c#L255~278](https://github.com/sysprog21/mado/blob/35ddf50ac4c7fe61b8b3c963f80dabceeb7783fe/backend/linux_input.c#L255)? 

Or we just assign NULL to `.poll`?
```c
/* Register the Linux DRM backend */

const twin_backend_t g_twin_backend = {
    .init = twin_drm_init,
    .poll = NULL,  /*without implementation*/
    .configure = twin_drm_configure,
    .exit = twin_drm_exit,
};
```

jserv replied:
@alanjian85, can you comment this?

jserv commented on README.md:
````diff
@@ -128,6 +131,14 @@ $ sudo usermod -a -G video $USERNAME
+$ sudo ./demo-drm
+```
+
+The DRM device can be assigened via the environment variable `DRI_CARD`.
````
`DRI_CARD` is not common for understanding. How about `DRMDEV`?

jserv reviewed (changes requested):
Rebase the latest `main` branch and adapt recent Linux virtual terminal changes.

jserv reviewed (changes requested):
Refine git commit messages:
1. Use `Close #60` instead of `Close sysprog21#60`
2. Always use third-person perspective: Avoid personal pronouns like "I", "you", and "we", focusing on an objective description of changes.
3. More explanation for using legacy DRM interface. Provide context for why the alternatives are not suitable.

huaxinliao commented on backend/drm.c:
```diff
@@ -0,0 +1,302 @@
+    if (drmModeAddFB(tx->drm_dri_fd, tx->width, tx->height, 24, 32,
+                     create_req.pitch, create_req.handle, &tx->fb_id) != 0) {
+        log_error("Cannot create framebubber");
+        return false;
```
I consider that this line should be changed to `goto bail_dumb`.
```c
bail_dumb:
    struct drm_mode_destroy_dumb destroy = { .handle = create_req.handle };
    ioctl(tx->drm_dri_fd, DRM_IOCTL_MODE_DESTROY_DUMB, &destroy);
    return false;
```

huaxinliao commented on backend/drm.c:
```diff
@@ -0,0 +1,302 @@
+    /* Create framebuffer object for the dumb-buffer */
+    if (drmModeAddFB(tx->drm_dri_fd, tx->width, tx->height, 24, 32,
+                     create_req.pitch, create_req.handle, &tx->fb_id) != 0) {
+        log_error("Cannot create framebubber");
```
Correct the typo in `framebubber`.

huaxinliao commented on backend/drm.c:
```diff
@@ -0,0 +1,302 @@
+    twin_screen_resize(SCREEN(ctx), width, height);
+}
+
+static void twin_drm_exit(twin_context_t *ctx)
```
I consider that `twin_drm_exit` should handle `ioctl(tx->drm_dri_fd, DRM_IOCTL_MODE_DESTROY_DUMB, &destroy)` to ensure allocated resources are properly released.

huaxinliao commented:
I ran the DRM backend on a VM with Ubuntu 24.04. However, I noticed that the screen froze.

https://github.com/user-attachments/assets/6c92de83-3e8b-450f-b544-6769070ed295

Bennctu commented:
> I ran the DRM backend on a VM with Ubuntu 24.04. However, I noticed that the screen froze.
> drm.mov

@huaxinliao Thank you for reporting this issue.

jserv commented:
For mado, a libdrm-free DRM backend is realistic because the use caseonly needs KMS + dumb buffer + CPU rendering, not GBM/EGL/GPU rendering. This pull request currently uses libdrm helpers such as `drmModeGetResources`, `drmModeGetConnector`, `drmModeAddFB`, `drmModeSetCrtc`, and `drmModeRmFB`, while already using raw ioctls for dumb-buffer creation/mapping.

The dumb-buffer path is appropriate: DRM dumb buffers are CPU-mappable scanout buffers, suitable for simple rendering but not complex GPU composition. `GETCONNECTOR` is intentionally a two-pass ioctl: first retrieve counts, then allocate arrays and retrieve modes/encoders; hotplug can race, so retry until counts stabilize.

Minimal ioctl-only framebuffer sequence is shown below:
```c
open("/dev/dri/card0", O_RDWR | O_CLOEXEC);

DRM_IOCTL_MODE_GETRESOURCES;
DRM_IOCTL_MODE_GETCONNECTOR;   /* two-pass */
DRM_IOCTL_MODE_GETCRTC;

DRM_IOCTL_MODE_CREATE_DUMB;
DRM_IOCTL_MODE_ADDFB2;
DRM_IOCTL_MODE_MAP_DUMB;
mmap();

DRM_IOCTL_MODE_SETCRTC;
```

Use legacy `SETCRTC` first, not atomic KMS. Atomic is cleaner long-term, but it requires property discovery and more code. For mado’s CPU-rendered full-screen backend, direct legacy KMS ioctls are the smallest maintainable step.

Bennctu committed:
Support Linux DRM backend

Implemented DRM-based framebuffer setup, including:
 - Opening the DRM device
 - Creating dumb buffers and mapping them to framebuffers
 - Setting mode and handling CRTC for connected display output

The legacy DRM ioctl interface is used as the current implementation
does not require atomic mode-setting or page flipping.

Atomic mode-setting prevents inconsistent states, supports rollback on
failures, and enables mode testing before committing changes.

Atomic page flipping synchronizes multiple planes within a VBLANK
interval, preventing tearing and improving display updates, especially
for power-efficient embedded controllers.

While the atomic API enhances KMS with property-based state management,
its complexity is unnecessary for now. The legacy interface remains
adequate, but future needs may justify adopting atomic mode-setting.

Close #60

Bennctu commented:
> I ran the DRM backend on a VM with Ubuntu 24.04. However, I noticed that the screen froze.
> drm.mov

I verified that the DRM backend runs successfully on Ubuntu 24.04 VM.

[Screencast from 2026年五月02日 00時56分24秒.webm](https://github.com/user-attachments/assets/f1788b25-07c5-42dd-81c2-6314db090f69)

huaxinliao commented:
I'm curious which part of the code fixed the VM issue.

shengwen-tw commented on backend/drm.c:
```diff
@@ -0,0 +1,533 @@
+DRM_PUT_SPAN_IMPL(16, ARGB32_TO_RGB565_PERLINE)
+DRM_PUT_SPAN_IMPL(24, ARGB32_TO_RGB888_PERLINE)
+DRM_PUT_SPAN_IMPL(32, ARGB32_TO_ARGB32_PERLINE)
+twin_context_t *twin_drm_init(int width maybe_unused, int height maybe_unused)
```
Please insert a new line here.

shengwen-tw commented on backend/drm.c:
```diff
@@ -0,0 +1,533 @@
+    } while (0)
+
+#define ARGB32_TO_ARGB32_PERLINE(dest, pixels, width) \
+    memcpy((uint32_t *) dest, pixels, width * 4)
```
Please wrap `dest`/ `pixels`/ `width` with parentheses in case they are passed as an expression, e.g. `ptr + offset`.

```
define ARGB32_TO_ARGB32_PERLINE(dest, pixels, width) \
    memcpy((uint32_t *)(dest), (pixels), (width) * sizeof(uint32_t))
```

shengwen-tw commented on backend/drm.c:
```diff
@@ -0,0 +1,533 @@
+/*
+ * Twin - A Tiny Window System
+ * Copyright (c) 2024 National Cheng Kung University, Taiwan
```
Please update the year to `2026`.

huaxinliao commented:
> Currently, using the DRM legacy interface is suitable for Mado's backend because we only need basic window display without advanced features.
> 
> <img alt="" width="500" src="https://private-user-images.githubusercontent.com/42055993/382581104-ec9f0ccb-e788-4f69-9896-85826c2d5c0a.JPG?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Nzc2NTczMTUsIm5iZiI6MTc3NzY1NzAxNSwicGF0aCI6Ii80MjA1NTk5My8zODI1ODExMDQtZWM5ZjBjY2ItZTc4OC00ZjY5LTk4OTYtODU4MjZjMmQ1YzBhLkpQRz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA1MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNTAxVDE3MzY1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFjY2JjZjJhOGQyZDJjODhmZmVjNWY1ZjFhZjI2MTc2MGVlMDczYzA1ZmQxMjlmYjM1ZWEyNmE0NmUxOWJhYmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRmpwZWcifQ.tUkhSgis3XaTcFJIRpxK7Y17CuXzM0X_tgWAc2lVLvk">

Hi @Bennctu, Also, does commit e4f93e9 show the same rendering result as this one? I’d like to learn more about this part as well. Thank you for your generous explanation!

jserv commented:
Thank @Bennctu for contributing!
